File: packages/java/anip-service/src/main/java/dev/anip/service/AuditAggregator.java
package dev.anip.service;

import java.time.Instant;
import java.util.*;

public class AuditAggregator {
    private final long windowSeconds;
    private final Map<WindowKey, Bucket> windows = new LinkedHashMap<>();

    public AuditAggregator(int windowSeconds) {
        this.windowSeconds = windowSeconds;
    }

    public synchronized void submit(Map<String, Object> event) {
        String actorKey = strOr(event, "actor_key", "anonymous");
        String capability = strOr(event, "capability", "_pre_auth");
        String failureType = strOr(event, "failure_type", "unknown");
        Instant ts = parseTimestamp(event);
        long epoch = ts.getEpochSecond() - (ts.getEpochSecond() % windowSeconds);
        var wk = new WindowKey(actorKey, capability, failureType, epoch);
        var bucket = windows.computeIfAbsent(wk, k -> {
            String detail = (String) event.get("detail");
            if (detail != null && detail.length() > 200) detail = detail.substring(0, 200);
            return new Bucket(ts, detail);
        });
        bucket.events.add(event);
        if (ts.isAfter(bucket.lastSeen)) bucket.lastSeen = ts;
    }

    public synchronized List<Object> flush(Instant now) {
        List<Object> results = new ArrayList<>();
        long nowEpoch = now.getEpochSecond();
        var it = windows.entrySet().iterator();
        while (it.hasNext()) {
            var e = it.next();
            var wk = e.getKey();
            var bucket = e.getValue();
            long windowEnd = wk.epoch + windowSeconds;
            if (windowEnd > nowEpoch) continue;
            String windowStart = Instant.ofEpochSecond(wk.epoch).toString();
            String windowEndStr = Instant.ofEpochSecond(windowEnd).toString();
            if (bucket.events.size() == 1) {
                results.add(bucket.events.get(0));
            } else {
                Map<String, Object> aggregated = new LinkedHashMap<>();
                aggregated.put("entry_type", "aggregated");
                aggregated.put("event_class", "repeated_low_value_denial");
                aggregated.put("retention_tier", "aggregate_only");
                aggregated.put("grouping_key", Map.of(
                    "actor_key", wk.actorKey, "capability", wk.capability, "failure_type", wk.failureType));
                aggregated.put("aggregation_window", Map.of("start", windowStart, "end", windowEndStr));
                aggregated.put("aggregation_count", bucket.events.size());
                aggregated.put("count", bucket.events.size());
                aggregated.put("first_seen", bucket.firstSeen.toString());
                aggregated.put("last_seen", bucket.lastSeen.toString());
                aggregated.put("representative_detail", bucket.representativeDetail);
                aggregated.put("capability", wk.capability);
                aggregated.put("failure_type", wk.failureType);
                aggregated.put("success", false);
                results.add(aggregated);
            }
            it.remove();
        }
        return results;
    }

    private static String strOr(Map<String, Object> m, String key, String def) {
        Object v = m.get(key);
        return v instanceof String s && !s.isEmpty() ? s : def;
    }

    private static Instant parseTimestamp(Map<String, Object> event) {
        Object ts = event.get("timestamp");
        if (ts instanceof String s) {
            try { return Instant.parse(s); } catch (Exception ignored) {}
        }
        return Instant.now();
    }

    private record WindowKey(String actorKey, String capability, String failureType, long epoch) {}

    private static class Bucket {
        final List<Map<String, Object>> events = new ArrayList<>();
        final Instant firstSeen;
        Instant lastSeen;
        final String representativeDetail;
        Bucket(Instant firstSeen, String detail) {
            this.firstSeen = firstSeen;
            this.lastSeen = firstSeen;
            this.representativeDetail = detail;
        }
    }
}

File: packages/csharp/src/Anip.Service/AuditAggregator.cs
namespace Anip.Service;

/// <summary>
/// Time-window bucketed aggregation for low-value denial events.
/// Implements SPEC section 6.9.
/// </summary>
public class AuditAggregator
{
    private readonly long _windowSeconds;
    private readonly object _lock = new();
    private readonly Dictionary<(string ActorKey, string Capability, string FailureType, long Epoch), Bucket> _windows = new();

    public AuditAggregator(int windowSeconds)
    {
        _windowSeconds = windowSeconds;
    }

    /// <summary>
    /// Adds an event to the aggregator.
    /// </summary>
    public void Submit(Dictionary<string, object?> eventData)
    {
        lock (_lock)
        {
            var actorKey = GetString(eventData, "actor_key") ?? "anonymous";
            var capability = GetString(eventData, "capability") ?? "_pre_auth";
            var failureType = GetString(eventData, "failure_type") ?? "unknown";

            var ts = ParseTimestamp(eventData);
            var epoch = ts.ToUnixTimeSeconds() - (ts.ToUnixTimeSeconds() % _windowSeconds);

            var key = (actorKey, capability, failureType, epoch);

            if (!_windows.TryGetValue(key, out var bucket))
            {
                var detail = GetString(eventData, "detail") ?? "";
                if (detail.Length > 200)
                    detail = detail[..200];

                bucket = new Bucket
                {
                    FirstSeen = ts,
                    LastSeen = ts,
                    RepresentativeDetail = detail,
                };
                _windows[key] = bucket;
            }

            bucket.Events.Add(eventData);
            if (ts > bucket.LastSeen)
                bucket.LastSeen = ts;
        }
    }

    /// <summary>
    /// Closes all windows whose end time is less than or equal to now and returns the results.
    /// Single-event buckets pass through as Dictionary. Multi-event buckets produce AggregatedEntry.
    /// </summary>
    public List<object> Flush(DateTimeOffset now)
    {
        lock (_lock)
        {
            var results = new List<object>();
            var nowUnix = now.ToUnixTimeSeconds();
            var toRemove = new List<(string, string, string, long)>();

            foreach (var (key, bucket) in _windows)
            {
                var windowEnd = key.Epoch + _windowSeconds;
                if (windowEnd > nowUnix)
                    continue; // window still open

                var windowStartStr = DateTimeOffset.FromUnixTimeSeconds(key.Epoch).UtcDateTime.ToString("o");
                var windowEndStr = DateTimeOffset.FromUnixTimeSeconds(windowEnd).UtcDateTime.ToString("o");

                if (bucket.Events.Count == 1)
                {
                    results.Add(bucket.Events[0]);
                }
                else
                {
                    results.Add(new AggregatedEntry
                    {
                        EventClass = "repeated_low_value_denial",
                        RetentionTier = "aggregate_only",
                        GroupingKey = new Dictionary<string, string>
                        {
                            ["actor_key"] = key.ActorKey,
                            ["capability"] = key.Capability,
                            ["failure_type"] = key.FailureType,
                        },
                        WindowStart = windowStartStr,
                        WindowEnd = windowEndStr,
                        Count = bucket.Events.Count,
                        FirstSeen = bucket.FirstSeen.UtcDateTime.ToString("o"),
                        LastSeen = bucket.LastSeen.UtcDateTime.ToString("o"),
                        RepresentativeDetail = bucket.RepresentativeDetail,
                    });
                }

                toRemove.Add(key);
            }

            foreach (var key in toRemove)
                _windows.Remove(key);

            return results;
        }
    }

    private static string? GetString(Dictionary<string, object?> dict, string key)
    {
        return dict.TryGetValue(key, out var val) ? val as string : null;
    }

    private static DateTimeOffset ParseTimestamp(Dictionary<string, object?> eventData)
    {
        if (eventData.TryGetValue("timestamp", out var tsObj) && tsObj is string ts)
        {
            if (DateTimeOffset.TryParse(ts, out var parsed))
                return parsed;
        }
        return DateTimeOffset.UtcNow;
    }

    private class Bucket
    {
        public List<Dictionary<string, object?>> Events { get; } = new();
        public DateTimeOffset FirstSeen { get; set; }
        public DateTimeOffset LastSeen { get; set; }
        public string RepresentativeDetail { get; set; } = "";
    }
}

/// <summary>
/// Emitted when a window closes with more than one event for a grouping key.
/// </summary>
public class AggregatedEntry
{
    public string EventClass { get; set; } = "";
    public string RetentionTier { get; set; } = "";
    public Dictionary<string, string> GroupingKey { get; set; } = new();
    public string WindowStart { get; set; } = "";
    public string WindowEnd { get; set; } = "";
    public int Count { get; set; }
    public string FirstSeen { get; set; } = "";
    public string LastSeen { get; set; } = "";
    public string RepresentativeDetail { get; set; } = "";

    /// <summary>
    /// Converts to a dictionary suitable for audit persistence.
    /// </summary>
    public Dictionary<string, object?> ToAuditDict()
    {
        return new Dictionary<string, object?>
        {
            ["entry_type"] = "aggregated",
            ["event_class"] = EventClass,
            ["retention_tier"] = RetentionTier,
            ["grouping_key"] = GroupingKey,
            ["aggregation_window"] = new Dictionary<string, string> { ["start"] = WindowStart, ["end"] = WindowEnd },
            ["aggregation_count"] = Count,
            ["count"] = Count,
            ["first_seen"] = FirstSeen,
            ["last_seen"] = LastSeen,
            ["representative_detail"] = RepresentativeDetail,
            ["capability"] = GroupingKey.TryGetValue("capability", out var cap) ? cap : null,
            ["failure_type"] = GroupingKey.TryGetValue("failure_type", out var ft) ? ft : null,
            ["success"] = false,
        };
    }
}
